tmpl/wiki/error-handling.md

```diff
@@ -0,0 +1,384 @@
+
+In theory, explicitly listing all possible errors and forcing callers to consider them should result in more reliable code.
+In practice, it results in callers ignoring all errors.
+
+The first problem is that there is no easy way to display an error to the user.
+If you look inside any `main.ml` file you will see code like this:
+
+```
+let t3 () =
+  console1 () >>= function
+  | `Error e -> fail (Failure "console1")
+  | `Ok console1 ->
+  block1 () >>= function
+  | `Error e -> fail (Failure "block1")
+  | `Ok block1 ->
+  return (`Ok (console1, block1))
+```
+
+Here, if the console or block devices return an error then we throw it away and report a generic error which just says which device failed, but not why.
+This approach to error handling can be seen throughout the Mirage code (grepping for `Error _` will turn up quite a few).
+Here's an example from `mirage-skeleton/block/unikernel.ml`, teaching new users how to redefine the bind operator to discard error details:
+
+```
+let ( >>= ) x f = x >>= function
+  | `Error _ -> fail (Failure "error")
+  | `Ok x -> f x
+```
+
+Another problem is that we force callers explicitly to handle errors that they would never want to handle.
+For example, in `BLOCK` we have:
+
+```
+module type BLOCK = sig
+  type error = [
+    | `Unknown of string (** an undiagnosed error *)
+    | `Unimplemented     (** operation not yet implemented in the code *)
+    | `Is_read_only      (** you cannot write to a read/only instance *)
+    | `Disconnected      (** the device has been previously disconnected *)
+  ]
+
+  val read:
+    t -> int64 -> page_aligned_buffer list ->
+    [ `Error of error | `Ok of unit ] io
+
+  ...
+```
+
+The implication is that callers of `read` should write code to handle each case.
+But how?
+
+```
+    B.read b start bufs >>= function
+    | `Error (`Unknown msg) -> fail (Failure msg)
+    | `Error `Unimplemented -> assert false
+    | `Error `Is_read_only -> assert false
+    | `Error `Disconnected -> fail (Failure "Disconnected")
+    | `Ok () ->
+```
+
+We might as well have just used exceptions and left the user in peace.
+
+```
+    B.read b start bufs >>= fun () ->
+```
+
+That way, we'd also get a stack trace showing what wasn't implemented, and we might get more information about why it was disconnected.
+
+A particular challenge for Mirage is that its types are abstract module types.
+Consider the `KV_RO` type:
+
+```
+module type KV_RO = sig
+  (** Static Key/value store. *)
+
+  type error = Unknown_key of string
+
+  val read:
+    t -> string -> int -> int ->
+    [ `Ok of page_aligned_buffer list | `Error of error ] io
+```
+
+For a simple in-memory store this error type allows us to represent all possible errors.
+However, if we want a store backed by a block device, there is no way to pass through errors such as the `Disconnected` one above.
+It is usually impossible to declare all the possible errors that an implementation of a generic interface might produce.
```

Comment:
That seems to be definitional defect that should be corrected. If you had a proper error type `('a, 'b) error = [````Ok of 'a | ````Error of 'b]`  and `Unknown_key` was a variant you could get through this.